networking/kvm.go

```diff
@@ -0,0 +1,214 @@
+// to allow communication with virtual machine created by lkvm tool
+// right now it only supports default "ptp" network type (other types ends with error)
+func kvmSetup(podRoot string, podID types.UUID, fps []ForwardedPort, privateNetList common.PrivateNetList, localConfig string) (*Networking, error) {
+	network := Networking{
+		podEnv: podEnv{
+			podRoot:      podRoot,
+			podID:        podID,
+			netsLoadList: privateNetList,
+			localConfig:  localConfig,
+		},
+	}
+	var e error
+	network.nets, e = network.loadNets()
+	if e != nil {
+		return nil, fmt.Errorf("error loading network definitions: %v", e)
+	}
+
+	for _, n := range network.nets {
+		switch n.conf.Type {
+		case "ptp":
+			link, err := setupTapDevice()
+			if err != nil {
+				return nil, err
+			}
+			ifName := link.Attrs().Name
+			n.runtime.IfName = ifName
+
+			err = kvmSetupNetAddressing(&network, n, ifName)
+			if err != nil {
+				return nil, err
+			}
+
+			// add address to host tap device
+			err = netlink.AddrAdd(
+				link,
+				&netlink.Addr{
+					IPNet: &net.IPNet{
+						IP:   n.runtime.HostIP,
+						Mask: net.IPMask(n.runtime.Mask),
+					},
+					Label: ifName,
+				})
+			if err != nil {
+				return nil, fmt.Errorf("cannot add address to host tap device %q: %v", ifName, err)
+			}
+
+			if n.conf.IPMasq {
+				chain := "CNI-" + n.conf.Name
+				if err = ip.SetupIPMasq(&net.IPNet{
+					IP:   n.runtime.IP,
+					Mask: net.IPMask(n.runtime.Mask),
+				}, chain); err != nil {
+					return nil, err
+				}
+			}
+		default:
+			return nil, fmt.Errorf("network %q have unsupported type: %q", n.conf.Name, n.conf.Type)
+		}
+	}
+	err := network.forwardPorts(fps, network.GetDefaultIP())
+	if err != nil {
+		return nil, err
+	}
+
+	return &network, nil
+}
+
+/*
+extend Networking struct with methods to clean up kvm specific network configurations
+*/
+
+// teardownKvmNets teardown every active networking from networking by
+// removing tuntap interface and releasing its ip from IPAM plugin
+func (n *Networking) teardownKvmNets() {
+	for _, an := range n.nets {
+		switch an.conf.Type {
+		case "ptp":
+			// remove tuntap interface
+			tuntap.RemovePersistentIface(an.runtime.IfName, tuntap.Tap)
+			// ugly hack again to directly call IPAM plugin to release IP
+			an.conf.Type = an.conf.IPAM.Type
+
+			_, err := n.execNetPlugin("DEL", &an, an.runtime.IfName)
+			if err != nil {
+				log.Printf("Error executing network plugin: %q", err)
+			}
+		default:
+			log.Printf("Unsupported network type: %q", an.conf.Type)
+		}
+	}
+}
+
+// kvmTeardown network teardown for kvm flavor based pods
+// similar to Networking.Teardown but without host namespaces
+func (n *Networking) kvmTeardown() {
+
+	if err := n.unforwardPorts(); err != nil {
+		log.Printf("Error removing forwarded ports (kvm): %v", err)
+	}
+	n.teardownKvmNets()
+
+}
+
+// NetParams exposes conf(NetConf)/runtime(NetInfo) data to stage1/init client
+type NetParams struct {
```

Comment:
I'm on the opposite side here: better to have many small types (and small interfaces), and not expose too much data, all that to prevent coupling (also called law of Demeter, a.k.a. principle of least knowledge or ISP).

We were thinking about using NetInfo for this, but this structure is dedicated to store runtime information and I need both (static and runtime configuration).

If you really want avoid new type, I can pass anonymous type containing both NetInfo and NetConf or as you said just make activeNet public. There is a third option to create an interface type declared close to data consumer (getKVMNetArgs function) which requires methods like HostIP()/GuestIP()/Mask() etc... and made this methods available in activeNet (we won't share a type, but just a behavior) - I need to prototype these three solutions to find 'good enough' one - I'll let you know.